UDFs (#15075)
cockroachdb/docs: v22.2/user-defined-functions.md

---
title: User-Defined Functions

> summary: A user-defined function is a named function defined at the database layer that can be called in queries and other contexts.

Reviewer: ```suggestion
summary: A user-defined function is a named function defined at the database level that can be called in queries and other contexts.
```

toc: true
key: sql-expressions.html
docs_area: reference.sql
---

> {% include_cached new-in.html version="v22.2" %} A user-defined function (UDF) is a named function defined at the database that can be called in queries and other contexts. CockroachDB supports invoking UDFs in `SELECT`, `FROM`, and `WHERE` clauses of [DML statements](sql-statements.html#data-manipulation-statements).

Reviewer: ```suggestion
{% include_cached new-in.html version="v22.2" %} A user-defined function (UDF) is a named function defined at the database level that can be called in queries and other contexts. CockroachDB supports invoking UDFs in `SELECT`, `FROM`, and `WHERE` clauses of [DML statements](sql-statements.html#data-manipulation-statements).
```

## Overview

The basic components of a user-defined function are a name, list of arguments, return type, volatility, language, and function body.

- An argument has a _mode_ and a _type_. CockroachDB supports the `IN` argument mode. The type can be a built-in type, [user-defined enum](enum.html), or implicit record type. CockroachDB **does not** support default values for arguments.
- The return type can be a built-in type, user-defined enum, implicit record type, or `VOID`. `VOID` indicates that there is no return type and `NULL` will always be returned. If the return type of the function is not `VOID`, the last statement of a UDF must be a `SELECT`.
- The [volatility](functions-and-operators.html#function-volatility) indicates whether the function has side effects. `VOLATILE` and `NOT LEAKPROOF` are the default.
  - Annotate a function with side effects with `VOLATILE`. This also prevents the [cost-based optimizer](cost-based-optimizer.html) from pre-evaluating the function.
  - A `STABLE` or `IMMUTABLE` function does not mutate data.

>   - `LEAKPROOF` indicates that a function has no side effects and that it communicates nothing that depends on its arguments besides the return value (i.e., it cannot throw an error that depends on the value of its arguments). You must precede `LEAKPROOF` with `IMMUTABLE` and only `IMMUTABLE` can be set to `LEAKPROOF`. `NOT LEAKPROOF` is allowed with any other volatility.

Reviewer: ```suggestion
  - `LEAKPROOF` indicates that a function has no side effects and that it communicates nothing that depends on its arguments besides the return value (i.e., it cannot throw an error that depends on the value of its arguments). You must precede `LEAKPROOF` with `IMMUTABLE`, and only `IMMUTABLE` can be set to `LEAKPROOF`. `NOT LEAKPROOF` is allowed with any other volatility.
```

- The language specifies the language of the function body. CockroachDB supports the language `SQL`.
- The function body:
  - Can reference arguments by name or by their ordinal in the function definition with the syntax `$1`.
  - Can be enclosed in a single line with single quotes `''` or multiple lines with `$$`.
  - Can reference tables.
  - Can reference only the `SELECT` statement.

## Examples

The following examples show how to create and invoke a simple UDF and view the definition of the UDF.

### Create a UDF

The following is a UDF that returns the sum of two integers:

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE FUNCTION add(a INT, b INT) RETURNS INT IMMUTABLE LEAKPROOF LANGUAGE SQL AS 'SELECT a + b';
~~~

> where:

Reviewer: ```suggestion
Where:
```

- name: `add`
- arguments: `a` of type `INT`, `b` of type `INT`
- return type: `INT`
- volatility: `IMMUTABLE LEAKPROOF`
- language: `SQL`
- function body: `'SELECT a + b'`

Alternatively, you could define this function as:

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE FUNCTION add(a INT, b INT) RETURNS INT IMMUTABLE LEAKPROOF LANGUAGE SQL AS 'SELECT $1 + $2';
~~~

> or

Reviewer: ```suggestion
or as:
```

> or as:

Reviewer: ```suggestion
Or as:
```

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE FUNCTION add(a INT, b INT) RETURNS INT LANGUAGE SQL AS $$
  SELECT a + b;
$$;
~~~

### Invoke a UDF

> You invoke the UDF like a [built-in function](functions-and-operators.html). For example:

Reviewer: ```suggestion
You invoke the UDF like a [built-in function](functions-and-operators.html):
```

{% include_cached copy-clipboard.html %}
~~~ sql
SELECT add(3,5) as sum;
~~~

~~~
  sum
-------
    8
(1 row)
~~~

### View a UDF definition

To view the `add` function definition, run:

{% include_cached copy-clipboard.html %}
~~~ sql
SHOW CREATE FUNCTION add;
~~~

> If you don't specify a schema for the function `add` when you create it, the default schema is `public`:

Reviewer: ```suggestion
If you do not specify a schema for the function `add` when you create it, the default schema is `public`:
```
